### PR TITLE
eth/downloader: Updated downloader tests to improve perf and reliability

### PR DESCRIPTION
**Pull request by rojotek**
eth/downloader: Updated use of Parallel and added some subtests to help isolate them.
Increased timeout in floodingTestPeer.RequestHeadersByNumber, so that it
doesn't accidently timeout and cause other tests to break.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: b464648640bda179fac278ef3bcba0f99559fb6e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: eth/downloader/downloader_test.go [hunk 1]
```diff
@@ -1665,12 +1672,26 @@ func testFakedSyncProgress(t *testing.T, protocol int, mode SyncMode) {
...
-func TestDeliverHeadersHang62(t *testing.T)      { testDeliverHeadersHang(t, 62, FullSync) }
-func TestDeliverHeadersHang63Full(t *testing.T)  { testDeliverHeadersHang(t, 63, FullSync) }
-func TestDeliverHeadersHang63Fast(t *testing.T)  { testDeliverHeadersHang(t, 63, FastSync) }
-func TestDeliverHeadersHang64Full(t *testing.T)  { testDeliverHeadersHang(t, 64, FullSync) }
-func TestDeliverHeadersHang64Fast(t *testing.T)  { testDeliverHeadersHang(t, 64, FastSync) }
-func TestDeliverHeadersHang64Light(t *testing.T) { testDeliverHeadersHang(t, 64, LightSync) }
+// We use data driven subtests to manage this so that it will be parallel on its own
+// and not with the other tests, avoiding intermittent failures.
+func TestDeliverHeadersHang(t *testing.T) {
+	testCases :=[]struct {
```
Comment (armaniferrante) [inline eth/downloader/downloader_test.go]: I think you forgot a space after `:=`

File: eth/downloader/downloader_test.go [hunk 2]
```diff
@@ -1742,12 +1762,28 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
...
 // Tests that if fast sync aborts in the critical section, it can restart a few
 // times before giving up.
-func TestFastCriticalRestartsFail63(t *testing.T) { testFastCriticalRestarts(t, 63, false) }
-func TestFastCriticalRestartsFail64(t *testing.T) { testFastCriticalRestarts(t, 64, false) }
-func TestFastCriticalRestartsCont63(t *testing.T) { testFastCriticalRestarts(t, 63, true) }
-func TestFastCriticalRestartsCont64(t *testing.T) { testFastCriticalRestarts(t, 64, true) }
+// We use data driven subtests to manage this so that it will be parallel on its own
+// and not with the other tests, avoiding intermittent failures.
+func TestFastCriticalRestarts(t *testing.T) {
+	testCases :=[]struct {
```
Comment (armaniferrante) [inline eth/downloader/downloader_test.go]: Another space missing here.

File: eth/downloader/downloader_test.go (hunk 1 above)
Comment (mexskican) [reply eth/downloader/downloader_test.go]: You should use [gofmt](https://golang.org/cmd/gofmt/) to avoid those kind of issues.

File: eth/downloader/downloader_test.go [hunk 3]
```diff
@@ -1665,12 +1672,26 @@ func testFakedSyncProgress(t *testing.T, protocol int, mode SyncMode) {
...
 // This test reproduces an issue where unexpected deliveries would
 // block indefinitely if they arrived at the right time.
-func TestDeliverHeadersHang62(t *testing.T)      { testDeliverHeadersHang(t, 62, FullSync) }
-func TestDeliverHeadersHang63Full(t *testing.T)  { testDeliverHeadersHang(t, 63, FullSync) }
-func TestDeliverHeadersHang63Fast(t *testing.T)  { testDeliverHeadersHang(t, 63, FastSync) }
-func TestDeliverHeadersHang64Full(t *testing.T)  { testDeliverHeadersHang(t, 64, FullSync) }
-func TestDeliverHeadersHang64Fast(t *testing.T)  { testDeliverHeadersHang(t, 64, FastSync) }
-func TestDeliverHeadersHang64Light(t *testing.T) { testDeliverHeadersHang(t, 64, LightSync) }
+// We use data driven subtests to manage this so that it will be parallel on its own
+// and not with the other tests, avoiding intermittent failures.
```
Comment (armaniferrante) [inline eth/downloader/downloader_test.go]: Why does running these tests synchronously cause intermittent failures?

Comment (rojotek) [reply eth/downloader/downloader_test.go]: I don't have a deep understanding, but suspect that a timeout in a different test was interacting badly with these.  There is still parallel tests being run, with all the header tests running in parallel, but not in parallel with the other tests.

**Comment by rojotek** [comment]
@armaniferrante @mexskican I've done the formatting fixes, and added a comment about the parallel, please let me know if you have any additional feedback.

**Comment by armaniferrante** [comment]
Looks good to me, though I'm uncertain about the two time related changes. Someone more familiar should also take a look as well.

File: eth/downloader/downloader_test.go [hunk 4]
```diff
@@ -1776,6 +1812,7 @@ func testFastCriticalRestarts(t *testing.T, protocol int, progress bool) {
 
 		// If it's the first failure, pivot should be locked => reenable all others to detect pivot changes
 		if i == 0 {
+			time.Sleep(150 * time.Millisecond) // Make sure no in-flight requests remain
```
Comment (fjl) [inline eth/downloader/downloader_test.go]: Why is this needed?

Comment (rojotek) [reply eth/downloader/downloader_test.go]: If you look at the lines of code just below, there is a sleep directly before failing.  I found that moving the sleep to just before the explicit fail was a better place to have it, and made the test more reliable.

Comment (rojotek) [reply eth/downloader/downloader_test.go]: @fjl Hi, just checking in to see if you saw my reply to your comment?
